onnxruntime/core/providers/cuda/nn/conv.cc

```diff
@@ -132,10 +136,37 @@ Status Conv<T, NHWC>::UpdateState(OpKernelContext* context, bool bias_expected)
       s_.cached_benchmark_results.clear();
     }
 
-    ORT_RETURN_IF_ERROR(conv_attrs_.ValidateInputShape(X->Shape(), W->Shape(), channels_last, channels_last));
+    // If we remove the contrib op NhwcConv we can rewrite this to be much simpler
+    // basically only strides of W will change and a transpose kernel is launched if NHWC is true
+    // currently we are only allowed to transpose the weight if node is part of kMSInternalNHWCDomain
+    if (transpose_weights_) {
+      // if the OP is registered in domain kMSInternalNHWCDomain the weights are not transposed beforehand.
+      auto nchw_strides = generateStrides(w_dims, false);
+      w_dims = {w_dims[0], w_dims[2], w_dims[3], w_dims[1]};
+      auto nhwc_strides = generateStrides(w_dims, true);
+      cudaStream_t compute_stream = Stream(context);
+      const T * w_data_nchw = W->Data<T>();
+      size_t weight_bytes = W->SizeInBytes();
+      // cudaMallocAsync(&w_data_nhwc_temp, weight_bytes, compute_stream);
+      IAllocatorUniquePtr<void> w_data_nhwc_temp = GetTransientScratchBuffer<void>(weight_bytes);
+      const TArray<int64_t> nchw_strides_arr = {};
+      const TArray<fast_divmod> nhwc_strides_arr = {};
+      auto status = TransposeImpl(
+        compute_stream, sizeof(T), w_dims.size(),
+        nchw_strides_arr, w_data_nchw, nhwc_strides_arr, w_data_nhwc_temp.get(), w_dims.size()
+      );
+
+      T * w_data_nchw_out = const_cast<T*>(W->Data<T>());
+      CUDA_RETURN_IF_ERROR(cudaMemcpyAsync(w_data_nchw_out ,w_data_nhwc_temp.get() , weight_bytes, cudaMemcpyDeviceToDevice, compute_stream));
+      CUDA_RETURN_IF_ERROR(cudaStreamSynchronize(compute_stream));
+      // cudaFreeAsync(w_data_nhwc_temp, compute_stream);
+    }
```

Comment:
@skottmckay Can you help me figure out if I actually have to transpose the weights ? All the tensors still "report" NCHW as their shape in the shape vector and I have also to found the attribute that specifies e.g. strides. `TensorPitches` for example does not seem to respect the underlying data layout ?

Reply:
I found this on XNN: 
https://github.com/microsoft/onnxruntime/blob/756eda2cc499cb56640d4b7320ebdaee2fd3751a/onnxruntime/core/providers/xnnpack/nn/conv.cc#L16-L44
Would that be the fitting method ? And if so judging by `tensor.Location()` the incoming tensor is already in CUDA memory while I would prefer having it in CPU memory and uploading myself to to NCHW->NHWC transition on CPU to save on peak memory usage on the GPU.

Reply:
Currently we don't transpose the weight input of Conv as it's {M, C/group, kH, kW} not NCHW. An argument can definitely be made that what matters is where the 'C' is regardless of whether it's 'C' or 'C/group'.

With the current setup, the PrePack is the place to do the Transpose of the weight input. PrePack is called during model loading as a one-off step. This approach also works better with a minimal build (used in scenarios where binary size is critical like mobile) as there's no constant folding available to actually change the layout of the weights during model loading, so if we inserted a Transpose node we'd be changing the weight layout every time we called Run. 

We could potentially change the setup to insert a Transpose for the weight input but would need to enable the constant folding optimizer (or something equivalent) in a minimal build (potentially non-trivial).

As for peak memory usage, the original buffer is released as soon as the pre-pack is done for that buffer.

Reply:
Thanks Scott, I went with prepackaged for now. It would still be valid if we decide to insert transpose nodes I guess so ...